// @magicblock-labs/soar-sdk/lib/state/achievement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AchievementAccount = void 0;
var AchievementAccount = (function () {
    function AchievementAccount(address, game, id, title, description, nftMeta, reward) {
        this.address = address;
        this.game = game;
        this.id = id;
        this.title = title;
        this.description = description;
        this.nftMeta = nftMeta;
        this.reward = reward;
    }
    AchievementAccount.fromIdlAccount = function (account, address) {
        return new AchievementAccount(address, account.game, account.id, account.title, account.description, account.nftMeta, account.reward);
    };
    AchievementAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            game: this.game.toBase58(),
            id: this.id.toString(),
            title: this.title,
            description: this.description,
            nftMeta: this.nftMeta.toBase58(),
            reward: this.reward !== null ? this.reward.toBase58() : this.reward,
        };
    };
    return AchievementAccount;
}());
exports.AchievementAccount = AchievementAccount;
