// @magicblock-labs/soar-sdk/lib/state/game.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Genre = exports.GameType = exports.GameAccount = void 0;
var soar_game_1 = require("../soar.game");
var GameAccount = (function () {
    function GameAccount(_address, account) {
        this.address = _address;
        this.meta = account.meta;
        this.achievementCount = account.achievementCount;
        this.leaderboardCount = account.leaderboardCount;
        this.auth = account.auth;
    }
    GameAccount.fromIdlAccount = function (account, address) {
        return new GameAccount(address, account);
    };
    GameAccount.prototype.client = function (soar) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, new soar_game_1.GameClient(soar, this.address, this)];
            });
        });
    };
    GameAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            meta: __assign(__assign({}, this.meta), { nftMeta: this.meta.nftMeta.toBase58() }),
            achievementCount: this.achievementCount.toString(),
            leaderboardCount: this.leaderboardCount.toString(),
            auth: this.auth.map(function (auth) { return auth.toBase58(); }),
        };
    };
    return GameAccount;
}());
exports.GameAccount = GameAccount;
var GameType;
(function (GameType) {
    GameType[GameType["Mobile"] = 0] = "Mobile";
    GameType[GameType["Desktop"] = 1] = "Desktop";
    GameType[GameType["Web"] = 2] = "Web";
    GameType[GameType["Unspecified"] = 255] = "Unspecified";
})(GameType || (exports.GameType = GameType = {}));
var Genre;
(function (Genre) {
    Genre[Genre["RPG"] = 0] = "RPG";
    Genre[Genre["MMO"] = 1] = "MMO";
    Genre[Genre["Action"] = 2] = "Action";
    Genre[Genre["Adventure"] = 3] = "Adventure";
    Genre[Genre["Puzzle"] = 4] = "Puzzle";
    Genre[Genre["Casual"] = 5] = "Casual";
    Genre[Genre["Unspecified"] = 255] = "Unspecified";
})(Genre || (exports.Genre = Genre = {}));
