// @magicblock-labs/soar-sdk/lib/state/leaderboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeaderBoardAccount = void 0;
var LeaderBoardAccount = (function () {
    function LeaderBoardAccount(address, id, game, description, nftMeta, decimals, minScore, maxScore, allowMultipleScores, topEntries) {
        this.address = address;
        this.id = id;
        this.game = game;
        this.description = description;
        this.nftMeta = nftMeta;
        this.decimals = decimals;
        this.minScore = minScore;
        this.maxScore = maxScore;
        this.allowMultipleScores = allowMultipleScores;
        this.topEntries = topEntries;
    }
    LeaderBoardAccount.fromIdlAccount = function (account, address) {
        return new LeaderBoardAccount(address, account.id, account.game, account.description, account.nftMeta, account.decimals, account.minScore, account.maxScore, account.allowMultipleScores, account.topEntries);
    };
    LeaderBoardAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            id: this.id.toString(),
            game: this.game.toBase58(),
            description: this.description,
            nftMeta: this.nftMeta.toBase58(),
            decimals: this.decimals,
            minScore: this.minScore.toString(),
            maxScore: this.maxScore.toString(),
            allowMultipleScores: this.allowMultipleScores,
            topEntries: this.topEntries ? this.topEntries.toBase58() : null,
        };
    };
    return LeaderBoardAccount;
}());
exports.LeaderBoardAccount = LeaderBoardAccount;
