// @magicblock-labs/soar-sdk/lib/state/leaderboardTopEntries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopEntriesAccount = void 0;
var TopEntriesAccount = (function () {
    function TopEntriesAccount(address, isAscending, topScores) {
        this.address = address;
        this.isAscending = isAscending;
        this.topScores = topScores;
    }
    TopEntriesAccount.fromIdlAccount = function (account, address) {
        return new TopEntriesAccount(address, account.isAscending, account.topScores);
    };
    TopEntriesAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            isAscending: this.isAscending,
            topScores: this.topScores.map(function (score) { return printLeaderboardScore(score); }),
        };
    };
    return TopEntriesAccount;
}());
exports.TopEntriesAccount = TopEntriesAccount;
var printLeaderboardScore = function (raw) {
    return {
        player: raw.player.toBase58(),
        entry: {
            score: raw.entry.score.toString(),
            timestamp: raw.entry.score.toString(),
        },
    };
};
