// @magicblock-labs/soar-sdk/lib/state/merged.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MergedAccount = void 0;
var MergedAccount = (function () {
    function MergedAccount(address, initiator, approvals, mergeComplete) {
        this.address = address;
        this.initiator = initiator;
        this.approvals = approvals;
        this.mergeComplete = mergeComplete;
    }
    MergedAccount.fromIdlAccount = function (account, address) {
        return new MergedAccount(address, account.initiator, account.approvals, account.mergeComplete);
    };
    MergedAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            initiator: this.initiator.toBase58(),
            others: this.approvals.map(function (approval) { return printMergeApproval(approval); }),
            mergeComplete: this.mergeComplete,
        };
    };
    return MergedAccount;
}());
exports.MergedAccount = MergedAccount;
var printMergeApproval = function (raw) {
    return {
        key: raw.key.toBase58(),
        approved: raw.approved,
    };
};
