// @magicblock-labs/soar-sdk/lib/state/playerAchievement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerAchievementAccount = void 0;
var PlayerAchievementAccount = (function () {
    function PlayerAchievementAccount(address, player, achievement, timestamp, unlocked, claimed) {
        this.address = address;
        this.player = player;
        this.achievement = achievement;
        this.timestamp = timestamp;
        this.unlocked = unlocked;
        this.claimed = claimed;
    }
    PlayerAchievementAccount.fromIdlAccount = function (account, address) {
        return new PlayerAchievementAccount(address, account.playerAccount, account.achievement, account.timestamp, account.unlocked, account.claimed);
    };
    PlayerAchievementAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            player: this.player.toBase58(),
            achievement: this.achievement.toBase58(),
            timestamp: this.timestamp.toString(),
            unlocked: this.unlocked,
            claimed: this.claimed,
        };
    };
    return PlayerAchievementAccount;
}());
exports.PlayerAchievementAccount = PlayerAchievementAccount;
