// @magicblock-labs/soar-sdk/lib/state/playerScoresList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerScoresListAccount = void 0;
var PlayerScoresListAccount = (function () {
    function PlayerScoresListAccount(address, playerAccount, leaderboard, allocCount, scores) {
        this.address = address;
        this.playerAccount = playerAccount;
        this.leaderboard = leaderboard;
        this.allocCount = allocCount;
        this.scores = scores;
    }
    PlayerScoresListAccount.fromIdlAccount = function (account, address) {
        return new PlayerScoresListAccount(address, account.playerAccount, account.leaderboard, account.allocCount, account.scores);
    };
    PlayerScoresListAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            playerInfo: this.playerAccount.toBase58(),
            leaderboard: this.leaderboard.toBase58(),
            allocCount: this.allocCount,
            scores: this.scores.map(function (score) { return printScoreEntry(score); }),
        };
    };
    return PlayerScoresListAccount;
}());
exports.PlayerScoresListAccount = PlayerScoresListAccount;
var printScoreEntry = function (entry) {
    return {
        score: entry.score.toString(),
        timestamp: entry.timestamp.toString(),
    };
};
