// @magicblock-labs/soar-sdk/lib/state/reward.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardAccount = void 0;
var RewardAccount = (function () {
    function RewardAccount(address, achievement, availableSpots, FungibleToken, NonFungibleToken) {
        this.address = address;
        this.achievement = achievement;
        this.availableSpots = availableSpots;
        this.FungibleToken = FungibleToken;
        this.NonFungibleToken = NonFungibleToken;
    }
    RewardAccount.fromIdlAccount = function (account, address) {
        return new RewardAccount(address, account.achievement, account.availableSpots, account.reward.fungibleToken, account.reward.nonFungibleToken);
    };
    RewardAccount.prototype.pretty = function () {
        return {
            address: this.address.toBase58(),
            achievement: this.achievement.toBase58(),
            availableSpots: this.availableSpots.toString(),
            FungibleToken: this.FungibleToken !== undefined
                ? {
                    mint: this.FungibleToken.mint.toBase58(),
                    account: this.FungibleToken.account.toBase58(),
                    amount: this.FungibleToken.amount.toString(),
                }
                : undefined,
            NonFungibleToken: this.NonFungibleToken !== undefined
                ? {
                    uri: this.NonFungibleToken.uri,
                    name: this.NonFungibleToken.name,
                    symbol: this.NonFungibleToken.symbol,
                    minted: this.NonFungibleToken.minted.toString(),
                    collection: this.NonFungibleToken.collection !== null
                        ? this.NonFungibleToken.collection.toBase58()
                        : null,
                }
                : undefined,
        };
    };
    return RewardAccount;
}());
exports.RewardAccount = RewardAccount;
